### PR TITLE
[DSCP-165] Disciplina faucet

### PR DESCRIPTION
**Pull request by Martoon-00**
Implements disciplina faucet. Made following templates of witness and educator (for Launcher modules) and student api (for server).

API tested, although I'm not sure I can check that money are actually occur on destination address, need to do that later when witness is completed

### CODE REVIEW

File: scripts/launch/node.sh [hunk 1]
```diff
@@ -53,11 +57,23 @@ witness_params="
...
 --db-path $tmp_files/witness.db
 --log-dir $tmp_files/logs
---witness-listen 127.0.0.1:4020
+--witness-listen $witness_web_addr
 --witness-keyfile $tmp_files/witness.key
 --witness-gen-key
 "
 
+# parameters for faucet
+faucet_params="
```
Comment (volhovm) [inline scripts/launch/node.sh]: It's better to write a separate `faucet.sh` script that will be used on the deployment.

File: faucet/src/Dscp/Faucet/Web/Logic.hs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+-- its first time.
+ensureFirstGift
+    :: (MonadIO m, MonadThrow m, HasCtx ctx m '[GiftedAddresses])
+    => Address -> m ()
+ensureFirstGift addr = do
+    giftedAddresses <- view (lensOf @GiftedAddresses)
+    prevVal <- atomically . modifyTVarS giftedAddresses $
+        at addr <<.= Just ()
+    when (isJust prevVal) $
+        throwM AddressAlreadyGifted
```
Comment (volhovm) [inline faucet/src/Dscp/Faucet/Web/Logic.hs]: How do these translate to user errors? Are they returned to the frontend?

File: faucet/src/Dscp/Faucet/Web/Error.hs [hunk 3]
```diff
@@ -49,3 +56,13 @@ toServantErr err = (toServantErrNoReason err){ errBody = encode err }
...
+
+---------------------------------------------------------------------------
+-- Other
+---------------------------------------------------------------------------
+
+data FaucetDecodeErrTag
+instance Reifies FaucetDecodeErrTag String where
+    reflect _ = decodeUtf8 $ encode InvalidFormat
+
+type DSON = SimpleJSON FaucetDecodeErrTag
```
Comment (volhovm) [inline faucet/src/Dscp/Faucet/Web/Error.hs]: Comment required ))

File: faucet/src/Dscp/Faucet/Web/Logic.hs (hunk 2 above)
Comment (Martoon-00) [reply faucet/src/Dscp/Faucet/Web/Logic.hs]: Yes, they are `show`n and passed to json

File: scripts/launch/node.sh (hunk 1 above)
Comment (Martoon-00) [reply scripts/launch/node.sh]: Lol, I couldn't get what's wrong with it

File: faucet/src/Dscp/Faucet/Variables.hs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    ) where
+
+import Control.Lens (makeLenses)
+
+import Dscp.Core
+
+-- | Keeps set of addresses which has already got money from faucet.
+type GiftedAddresses = TVar (Set Address)
+
+-- | Provides sequencial transaction submission.
```
[the remark this answers is not in the document]
Comment (Martoon-00) [reply faucet/src/Dscp/Faucet/Variables.hs]: Lol, I couldn't get what's wrong with it

File: witness/src/Dscp/Witness/Web/Logic.hs [hunk 5]
```diff
@@ -57,10 +60,18 @@ getAccountState addr = do
...
--- | Submit transaction.
+-- | Applies transaction everywhere.
 submitUserTx
-    :: (MonadIO m, MonadThrow m, MonadReader ctx m, HasLens' ctx MempoolVar)
+    :: (BasicWorkMode m, MonadReader ctx m, HasLens' ctx MempoolVar)
     => TxWitnessed -> m ()
 submitUserTx tw = do
     mempool <- view (lensOf @MempoolVar)
     addTxToMempool mempool (GMoneyTxWitnessed tw)
+    -- TODO: also send into network
```
[the remark this answers is not in the document]
Comment (Martoon-00) [reply witness/src/Dscp/Witness/Web/Logic.hs]: I'll do that when mempool / networking would be ready, need to test transaction making anyway.
So transaction flow is the following: `request -> queue -> listener (also send tx into network) -> mempool -> block`?

[the remark this answers is not in the document]
Comment (volhovm) [reply witness/src/Dscp/Witness/Web/Logic.hs]: Yes, the flow is correct.x

**Comment by Martoon-00** [comment]
Ready for review.

File: core/src/Dscp/Core/Genesis.hs [hunk 6]
```diff
@@ -59,7 +64,11 @@ distrToMap (Just addrs) (GDEqual cTotal) =
     in if sum (map (unCoin . snd) mapping) /= unCoin cTotal
        then error "distrToMap: equal summing failed"
        else GenAddressMap $ Map.fromList mapping
-distrToMap _ _ = error "distrToMap: param combination is invalid"
+distrElemToMap _ _ = error "distrToMap: param combination is invalid"
+
+distrToMap :: Maybe (NonEmpty Address) -> GenesisDistribution -> GenAddressMap
+distrToMap maddrs (GenesisDistribution distrs) =
+    fold $ fmap (distrElemToMap maddrs) distrs
```
Comment (volhovm) [inline core/src/Dscp/Core/Genesis.hs]: How exactly does it work? I would _expect_ it to sum coins if key collisions appear.

**Pull request merged**